src/arithmetic/algebraic_expression/algebraic_expression_optimization.c

```diff
@@ -24,20 +25,20 @@ static void _AlgebraicExpression_CollectOperands(AlgebraicExpression *root,
 	switch(root->type) {
 	case AL_OPERAND:
 		*operands = array_append(*operands, AlgebraicExpression_Clone(root));
+		root->operand.should_free = false; // The caller is the new owner of this operand.
 		break;
 	case AL_OPERATION:
 		switch(root->operation.op) {
-		case AL_EXP_TRANSPOSE:
-			// Transpose is considered as an operand.
-			*operands = array_append(*operands, AlgebraicExpression_Clone(root));
-			break;
 		case AL_EXP_ADD:
 		case AL_EXP_MUL:
 			child_count = AlgebraicExpression_ChildCount(root);
 			for(uint i = 0; i < child_count; i++) {
 				_AlgebraicExpression_CollectOperands(CHILD_AT(root, i), operands);
 			}
 			break;
+		case AL_EXP_TRANSPOSE:
+			assert("Encountered transpose operation node later than expected" && false);
+			break;
```

Comment:
Please revert, just to stay generic.
We shouldn't make too many assumption on call sequence.